rascaline/src/math/hyp2f1.rs

```diff
@@ -0,0 +1,292 @@
+                    }
+                    hw = f1;
+                }
+                hf = f0 * c0 + f1 * c1;
+            }
+        } else {
+            let ga = gamma(a);
+            let gb = gamma(b);
+            let gc = gamma(c);
+            let gca = gamma(c - a);
+            let gcb = gamma(c - b);
+            let gcab = gamma(c - a - b);
+            let gabc = gamma(a + b - c);
+            let c0 = gc * gcab / (gca * gcb);
+            let c1 = gc * gabc / (ga * gb) * (1.0 - x).powf(c - a - b);
+            let mut hf = 0.;
+            let mut r0 = c0;
+            let mut r1 = c1;
+            for k in (1..=250).map(|k| k as f64) {
+                r0 *= (a + k - 1.) * (b + k - 1.0) / (k * (a + b - c + k)) * (1.0 - x);
+                r1 *= (c - a + k - 1.) * (c - b + k - 1.0) / (k * (c - a - b + k)) * (1.0 - x);
+                hf += r0 + r1;
+                if (hf - hw).abs() < (hf).abs() * eps {
+                    break;
+                }
+                hw = hf;
+            }
+            hf += c0 + c1;
+        }
+    } else {
+        let mut a0 = 1.;
+        if c > a && c < a * 2.0 && c > b && c < b * 2.0 {
+            a0 = (1.0 - x).powf(c - a - b);
+            a = c - a;
+            b = c - b;
+        }
+        let mut hf = 1.;
+        let mut term = 1.;
+        for k in (1..=250).map(|k| k as f64) {
+            term = term * (a + k - 1.) * (b + k - 1.) / (k * (c + k - 1.)) * x;
+            hf += term;
+            if (hf - hw).abs() <= (hf).abs() * eps {
+                break;
+            }
+            hw = hf;
+        }
+        hf *= a0;
+    }
+
+    if x1 < 0.0 {
+        let c0 = 1.0 / (1. - x1).powf(aa);
+        hf *= c0;
+    }
+
+    return hf;
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use approx::assert_relative_eq;
+
+    #[test]
+    fn test_hyp2f1() {
+    // These  values (a, b, c, x, expected) were computed using
+    // scipy.special.hyp2f1
+    let reference = [
+        (1.0, 2.0, 3.0, 0.0, 1.0),
+        (0.0, 1.0, 0.0, 0.0, 1.0),
+        (0.0, 0.0, 0.0, 0.0, 1.0),
+        //(0.1, 1.0, 2.0, -5.0, 0.8923895138816937),
+        (0.0, 0.0, 0.0, -1.0, 1.0),
+        (1.0, 1.0, 0.0, -1.0, -std::f64::INFINITY),
+        (-2.0, -2.0, -3.0, 0.5, std::f64::INFINITY),
+        (-7.2, -0.5, -4.5, 1e-05, 0.999991999964571),
+        (1.0, 2.0, -3.0, -1.0, std::f64::INFINITY),
+        (1.0, 0.5, -4.5, -20.0, 0.44674682523931875),
+        (-1.0, -2.0, 3.0, 0.0, 1.0),
+        (2.0, -2.0, -3.0, 0.5, std::f64::INFINITY),
+        (2.0, -3.0, -2.0, -1.0, std::f64::INFINITY),
+        (-7.2, 0.5, -4.5, 1e-05, 1.0000080001062877),
+        (-1.0, 2.0, 3.0, 0.0, 1.0),
+        (-1.0, -2.0, 4.0, -1.0, 0.5),
+        (0.0, -6.0, -4.0, -2.0, std::f64::INFINITY),
+        (0.0, -3.0, -4.0, -1.0, std::f64::INFINITY),
+        (1.0, -6.0, -4.0, -1.0, std::f64::INFINITY),
+        (1.0, -0.5, -4.5, -5.0, 0.46975635247240194),
+        (-7.2, -0.5, -4.5, -1.0, 19.915882647295806)
```

Comment:
I would also add some more extreme points, with large a, b, c, z values (50% larger that what we intend to use now)